Generated Haskell code has redundant matches triggering GHC warnings

Generated Haskell code produces tons of `overlapping-patterns` warning because we have a catch-all in every alternative:
``` 
warning: [-Woverlapping-patterns]
    Pattern match is redundant
    In a case alternative: _ -> ...
      |
12242 |       _ -> MAlonzo.RTE.mazUnreachableError
      |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Suggested fix: Put
```
{-# OPTIONS_GHC -Wno-overlapping-patterns -#}
```
into the generated Haskell files.

Note that we do rely on the `-Wincomplete-patterns` for checking that Agda data types bound to Haskell types are not leaving out any constructors:
```agda
data Bool : Set where
  true : Bool

{-# COMPILE GHC Bool = data Bool (True) #-}
```
fails with
```
MAlonzo/Code/Foo.hs:21:5: error: [-Wincomplete-patterns, -Werror=incomplete-patterns]
    Pattern match(es) are non-exhaustive
    In a case alternative: Patterns of type ‘Bool’ not matched: False
   |
21 |   = case x of
   |     ^^^^^^^^^...

```

--- a/Makefile
+++ b/Makefile
@@ -524,9 +524,14 @@ compiler-test :
 	@$(call decorate, "Compiler tests", \
 		AGDA_BIN=$(AGDA_BIN) $(AGDA_TESTS_BIN) $(AGDA_TESTS_OPTIONS) --regex-include all/Compiler --regex-exclude AllStdLib)
 
+.PHONY : ghc-compiler-test ##
+ghc-compiler-test :
+	@$(call decorate, "GHC Compiler tests", \
+		AGDA_BIN=$(AGDA_BIN) $(AGDA_TESTS_BIN) $(AGDA_TESTS_OPTIONS) --regex-include all/Compiler/MAlonzo_Lazy --regex-exclude AllStdLib)
+
 .PHONY : js-compiler-test ##
 js-compiler-test :
-	@$(call decorate, "Compiler tests", \
+	@$(call decorate, "JS Compiler tests", \
 		AGDA_BIN=$(AGDA_BIN) $(AGDA_TESTS_BIN) $(AGDA_TESTS_OPTIONS) --regex-include all/Compiler/JS_MinifiedOptimized --regex-exclude AllStdLib)
 
 .PHONY : std-lib-compiler-test ##

--- a/src/full/Agda/Compiler/MAlonzo/Compiler.hs
+++ b/src/full/Agda/Compiler/MAlonzo/Compiler.hs
@@ -1235,9 +1235,8 @@ writeModule (HS.Module m ps imp ds) = do
   liftIO $ UTF8.writeFile out $ (++ "\n") $ prettyPrint $
     -- TODO: It might make sense to skip bang patterns for the unused
     -- arguments of the "non-stripped" functions.
-    (if strict then makeStrict else id) $
+    applyWhen strict makeStrict $
     HS.Module m (p : ps) imp ds
-  where
 
 outFileAndDir :: MonadGHCIO m => HS.ModuleName -> m (FilePath, FilePath)
 outFileAndDir m = do
@@ -1262,9 +1261,7 @@ callGHC = do
   opts    <- askGhcOpts
   hsmod   <- prettyPrint <$> curHsMod
   agdaMod <- curAgdaMod
-  let outputName = case mnameToList agdaMod of
-        [] -> __IMPOSSIBLE__
-        m:ms -> last1 m ms
+  let outputName = lastWithDefault __IMPOSSIBLE__ $ mnameToList agdaMod
   (mdir, fp) <- curOutFileAndDir
   let ghcopts = optGhcFlags opts
 

--- a/src/full/Agda/Compiler/MAlonzo/Compiler.hs
+++ b/src/full/Agda/Compiler/MAlonzo/Compiler.hs
@@ -1233,11 +1233,22 @@ writeModule (HS.Module m ps imp ds) = do
             , "PatternSynonyms"
             , "OverloadedStrings"
             ]
+  let ghcOptions =
+        List.map HS.OtherPragma
+          [ ""  -- to separate from LANGUAGE pragmas
+          , "{-# OPTIONS_GHC -Wno-overlapping-patterns #-}"
+              -- Andreas, 2022-01-26, issue #5758:
+              -- Place this in generated file rather than
+              -- passing it only when calling GHC from within Agda.
+              -- This will silence the warning for the Agda-generated .hs
+              -- files while it can be on for other .hs files in the same
+              -- project.  (E.g., when using cabal/stack to compile.)
+          ]
   liftIO $ UTF8.writeFile out $ (++ "\n") $ prettyPrint $
     -- TODO: It might make sense to skip bang patterns for the unused
     -- arguments of the "non-stripped" functions.
     applyWhen strict makeStrict $
-    HS.Module m (concat [languagePragmas, ps]) imp ds
+    HS.Module m (concat [languagePragmas, ghcOptions, ps]) imp ds
 
 outFileAndDir :: MonadGHCIO m => HS.ModuleName -> m (FilePath, FilePath)
 outFileAndDir m = do
@@ -1285,7 +1296,6 @@ callGHC = do
         [ fp
         , "--make"
         , "-fwarn-incomplete-patterns"
-        , "-fno-warn-overlapping-patterns"
         ]
       args     = overridableArgs ++ ghcopts ++ otherArgs
 

--- a/src/full/Agda/Compiler/MAlonzo/Compiler.hs
+++ b/src/full/Agda/Compiler/MAlonzo/Compiler.hs
@@ -70,6 +70,7 @@ import Agda.Utils.List
 import Agda.Utils.Maybe
 import Agda.Utils.Monad
 import Agda.Utils.Pretty (prettyShow, render)
+import Agda.Utils.Singleton
 import qualified Agda.Utils.IO.UTF8 as UTF8
 import Agda.Utils.String
 
@@ -1212,7 +1213,16 @@ writeModule (HS.Module m ps imp ds) = do
   -- Note that GHC assumes that sources use ASCII or UTF-8.
   out <- snd <$> outFileAndDir m
   strict <- optGhcStrict <$> askGhcOpts
-  let p = HS.LanguagePragma $ List.map HS.Ident $
+  let languagePragmas =
+        List.map (HS.LanguagePragma . singleton . HS.Ident) $
+          List.sort $
+            [ "QualifiedDo" | strict ] ++
+                -- If --ghc-strict is used, then the language extension
+                -- QualifiedDo is activated. At the time of writing no
+                -- code is generated that depends on this extension
+                -- (except for the pragmas), but --ghc-strict is broken
+                -- with at least some versions of GHC prior to version 9,
+                -- and QualifiedDo was introduced with GHC 9.
             [ "BangPatterns"
             , "EmptyDataDecls"
             , "EmptyCase"
@@ -1222,21 +1232,12 @@ writeModule (HS.Module m ps imp ds) = do
             , "RankNTypes"
             , "PatternSynonyms"
             , "OverloadedStrings"
-            ] ++
-            -- If --ghc-strict is used, then the language extension
-            -- QualifiedDo is activated. At the time of writing no
-            -- code is generated that depends on this extension
-            -- (except for the pragmas), but --ghc-strict is broken
-            -- with at least some versions of GHC prior to version 9,
-            -- and QualifiedDo was introduced with GHC 9.
-            if strict
-            then ["QualifiedDo"]
-            else []
+            ]
   liftIO $ UTF8.writeFile out $ (++ "\n") $ prettyPrint $
     -- TODO: It might make sense to skip bang patterns for the unused
     -- arguments of the "non-stripped" functions.
     applyWhen strict makeStrict $
-    HS.Module m (p : ps) imp ds
+    HS.Module m (concat [languagePragmas, ps]) imp ds
 
 outFileAndDir :: MonadGHCIO m => HS.ModuleName -> m (FilePath, FilePath)
 outFileAndDir m = do

--- a/src/full/Agda/Compiler/MAlonzo/Pretty.hs
+++ b/src/full/Agda/Compiler/MAlonzo/Pretty.hs
@@ -17,12 +17,15 @@ prettyPrint = show . pretty
 
 instance Pretty HS.Module where
   pretty (HS.Module m pragmas imps decls) =
-    vcat [ vcat $ map pretty pragmas
-         , "module" <+> pretty m <+> "where"
-         , ""
-         , vcat $ map pretty imps
-         , ""
-         , vcat $ map pretty decls ]
+    vcat $ concat
+      [ map pretty pragmas
+      , [ "" | not $ null pragmas ]
+      , [ "module" <+> pretty m <+> "where" ]
+      , [ "" ]
+      , map pretty imps
+      , [ "" ]
+      , map pretty decls
+      ]
 
 instance Pretty HS.ModulePragma where
   pretty (HS.LanguagePragma ps) =

